### PR TITLE
Preprocess keywords and use sentence score to determine tracker lifetime

### PR DESCRIPTION
**Pull request by benfrancis314**
Apologies for putting two things in one PR; the lifetime estimation was just directly related to the new code for the preprocessing and was pretty simple to implement. 

Also note that this is going to likely cause some inaccuracy in the correctness of the timer, though it shouldn't be very bad. Added a TODO for that

### CODE REVIEW

File: extension/content_script.js [hunk 1]
```diff
@@ -502,10 +469,10 @@ function oneTimeSetup() {
...
 	doc = new Doc(readableDomEls);
 	// If page is not readable, stop setting up the rest of the app.
-	if (doc.sentences.length === 0) {
-		debug("Stopping app init because page is not readable");
-		return;
-	}
+	// if (doc.sentences.length === 0) {
+	// 	debug("Stopping app init because page is not readable");
+	// 	return;
+	// }
```
Comment (benfrancis314) [inline extension/content_script.js:474]: Temporarily commented out; doesn't work bc doc isn't ready (async problems) by the time this is called

File: extension/lib/doc.js [hunk 2]
```diff
@@ -125,21 +134,79 @@ class Doc {
                     element: "readerease-sentence"
                 });
                 this.sentenceEls.push($("." + sentenceClassName));
+                // Set keywords and score for each sentence
+                // Note: Putting this outside setSentenceKeywordsAndScore so it doesn't have to carry over as much text 
+                this.setSentenceKeywordsAndScore(container, containerText, start, end, sentenceId);
             }
+            this.num_words_per_sentence = this.calcNumWordsPerSentence(this.sentences);
+            this.num_words_per_sentence_suffix_sum = suffix_sum(this.num_words_per_sentence);
```
Comment (benfrancis314) [inline extension/lib/doc.js:142]: Moved in here due to async needs, this.sentences isn't ready if this is called in the constructor

File: extension/content_script.js [hunk 3]
```diff
@@ -263,6 +268,7 @@ function unhighlightEverything() {
 	if (tracker.isTracking()) {
 		let sentenceId = tracker.getSentenceId();
 		doc.getSentenceEls(tracker.getSentenceId()).removeClass(trackerStyle.getSentenceStyle());
+		doc.getSentenceKeywordsEls(tracker.getSentenceId()).removeClass(trackerStyle.getKeywordStyle());
 	}
 	$("." + keywordClass).unmark();
```
Comment (sjonany) [inline extension/content_script.js]: We probably don't need this anymore?
-- Also maybe ctrl - F all usages of keywordClass and make sure they all use the preprocessed keywordEls instead.

File: extension/content_script.js [hunk 4]
```diff
@@ -274,20 +280,17 @@ Highlight portion pointed to by tracker, and unhighlight previous sentence (if n
...
 	doc.getSentenceEls(sentenceId).addClass(sentenceStyle);
-
-	// Highlight they keywords.
-	let sentencePtr = doc.getSentence(sentenceId);
-	let container = doc.getContainer(sentencePtr.containerId);
-	let start = sentencePtr.start;
-	let end = sentencePtr.end;
-	highlightKeyWords(container, start, end);
+	// Highlight the keywords. 
+	doc.getSentenceKeywordsEls(sentenceId).addClass(keywordStyle)
```
Comment (sjonany) [inline extension/content_script.js]: aaaah looks so nice now :)

-- and semicolon

File: extension/lib/doc.js [hunk 5]
```diff
@@ -125,21 +134,80 @@ class Doc {
...
             if (this.sentences.length > MAX_NUM_SENTENCE) {
                 alert("Sorry, we don't support big documents yet :(");
                 this.containers = [];
                 this.sentences = [];
                 this.containerIdToFirstSentenceId = [];
                 break;
             }
+            this.containerIdToFirstSentenceId.push(this.sentences.length);
```
Comment (sjonany) [inline extension/lib/doc.js]: Why is this moved to the bottom, instead of letting it stay on the first few lines of the for loop?

** also not sure if we need containers anymore, but that's a separate refactoring cl

File: extension/lib/doc.js [hunk 6]
```diff
@@ -285,23 +375,24 @@ class Doc {
...
-        let documentFreqValue = documentFreq[word]
-        if (!documentFreqValue) {
-            keywords.add(word.toLowerCase()); // Word not encountered before -> Add to keywords
-            continue;
-        }
-        let relDocumentFreqValue = documentFreqValue / num_documents // Get relative document frequency
-        let tfIdfScore = relTermFreqValue / relDocumentFreqValue; // Score = TF/DF (hence TF-IDF)
-        if (tfIdfScore > keywordFilter) { // Filter by score
-            keywords.add(word.toLowerCase());
+   setKeyWords(termFreq, documentFreq, num_documents, cb) {
```
Comment (sjonany) [inline extension/lib/doc.js]: why have cb as parameter is this code is not async?
- remove cb() as param
- then caller just do setKeywords(); cb();

**Review by sjonany — CHANGES_REQUESTED** [review]

**Comment by sjonany** [comment]
Vvvv happy w/ the keyword preprocessing!
As for the timer calculation -- might want to prioritize the suffix sum fix. should be just a few lines cos u have suffix_sum function. but ok for later cl

File: extension/content_script.js (hunk 3 above)
Comment (benfrancis314) [reply extension/content_script.js]: 👍

File: extension/content_script.js (hunk 4 above)
Comment (benfrancis314) [reply extension/content_script.js]: Hahaha

File: extension/lib/doc.js (hunk 6 above)
Comment (benfrancis314) [reply extension/lib/doc.js]: Vestigial bug

Thanks!

File: extension/lib/doc.js (hunk 5 above)
Comment (benfrancis314) [reply extension/lib/doc.js]: Uncertain; moved back

**Review by sjonany — APPROVED** [review]

**Comment by sjonany** [comment]
LGTM
